app/lua/lflash.c

```diff
@@ -270,5 +298,261 @@ LUAI_FUNC int luaN_index (lua_State *L) {
+  lua_pushnil(out->L);   /* can't use it on a cpcall anyway */
+  lua_error(out->L);
+}
+
+/*
+ * uzlib_inflate does a stream inflate on an RFC 1951 encoded data stream.
+ * It uses three application-specific CBs passed in the call to do the work:
+ *
+ * -  get_byte()     CB to return next byte in input stream
+ * -  put_byte()     CB to output byte to output buffer
+ * -  recall_byte()  CB to output byte to retrieve a historic byte from
+ *                   the output buffer.
+ *
+ *  Note that put_byte() also triggers secondary CBs to do further processing.
+ */
+uint8_t get_byte (void) {
+  if (--in->left < 0) {
+    /* Read next input block */
+    int remaining = in->len - in->bytesRead;
+    int wanted    = remaining >= READ_BLOCKSIZE ? READ_BLOCKSIZE : remaining;
+
+    if (vfs_read(in->fd, in->block, wanted) != wanted)
+      flash_error("read error on LFS image file");
+      
+    system_soft_wdt_feed();
+
+    in->bytesRead += wanted;
+    in->inPtr      = in->block;
+    in->left       = wanted-1;
+  }
+  return *in->inPtr++;
+}
+
+
+void put_byte (uint8_t value) {
+  int offset = out->ndx % WRITE_BLOCKSIZE;  /* counts from 0 */
+
+  out->block[0]->byte[offset++] = value;
+  out->ndx++;
+
+  if (offset == WRITE_BLOCKSIZE || out->ndx == out->len) {
+    if (out->fullBlkCB)
+      out->fullBlkCB();
+    /* circular shift the block pointers (redundant on last block, but so what) */
+    outBlock *nextBlock  = out->block[WRITE_BLOCKS - 1];
+    memmove(out->block+1, out->block, (WRITE_BLOCKS-1)*sizeof(void*));
+    out->block[0] = nextBlock ;
+  }
+}
+
+
+uint8_t recall_byte (uint offset) {
+  if(offset > DICTIONARY_WINDOW || offset >= out->ndx)
+    flash_error("invalid dictionary offset on inflate");
+  /* ndx starts at 1. Need relative to 0 */
+  uint n   = out->ndx - offset;
+  uint pos = n % WRITE_BLOCKSIZE;
+  uint blockNo = out->ndx / WRITE_BLOCKSIZE - n  / WRITE_BLOCKSIZE;
+  return out->block[blockNo]->byte[pos];
+}
+
+/*
+ * On the first pass the break index is set to call this process at the end
+ * of each completed output buffer.
+ *  -  On the first call, the Flash Header is checked.
+ *  -  On each call the CRC is rolled up for that buffer.
+ *  -  Once the flags array is in-buffer this is also captured.
+ * This logic is slightly complicated by the last buffer is typically short.
+ */
+int procFirstPass (void) {
+  int len = (out->ndx % WRITE_BLOCKSIZE) ?
+               out->ndx % WRITE_BLOCKSIZE : WRITE_BLOCKSIZE;
+  if (out->ndx <= WRITE_BLOCKSIZE) {
+    uint32_t fl;
+    /* Process the flash header and cache the FlashHeader fields we need */
+    FlashHeader *fh = cast(FlashHeader *, out->block[0]);
+    out->flashLen   = fh->flash_size;                         /* in bytes */
+    out->flagsLen   = (out->len-fh->flash_size)/WORDSIZE;     /* in words */
+    out->flash_sig  = fh->flash_sig;
+
+    if ((fh->flash_sig & ~FLASH_SIG_ABSOLUTE) != FLASH_SIG)
+      flash_error("incorrect LFS header signature");
```

Comment:
I would really like to have different messages here for absolute addresses and integer/float missmatch and also stating this explicitly.

Reply:
My urgency was to get an error returned.  You're right: maybe we should split these cases down.  Another change that I thought of making was to up and internal count in the flash_sig every time any format change is made to the LFS internal format.  Yes, it should be obvious that you use the `luac.cross` from the same build as the firmware, but I can see this as a bearpit that developers might fall into.

Reply:
Yes, especially when there will be an option to download prebuilt luac.cross binaries.
Maybe the version command then should also tell the user what (internal) version it is building.
And the node.flashreload() also of course.